File: java/api/src/main/java/org/dressdiscover/api/models/object/ObjectId.java
package org.dressdiscover.api.models.object;

import static com.google.common.base.Preconditions.checkNotNull;

import org.dressdiscover.api.models.collection.CollectionId;
import org.dressdiscover.api.models.collection.InvalidCollectionIdException;
import org.dressdiscover.api.models.institution.InstitutionId;
import org.thryft.waf.api.models.StringModelId;

public final class ObjectId extends StringModelId {
    public static ObjectId parse(final String value) throws InvalidObjectIdException {
        if (value.isEmpty()) {
            throw InvalidObjectIdException.create("empty string");
        }
        final int lastSep = value.lastIndexOf('/');
        if (lastSep == -1 || lastSep == value.length() - 1) {
            throw InvalidObjectIdException.create("malformed object ID: " + value);
        }
        CollectionId collectionId;
        try {
            collectionId = CollectionId.parse(value.substring(0, lastSep));
        } catch (final InvalidCollectionIdException e) {
            throw InvalidObjectIdException.create(e.toString());
        }
        final String unqualifiedObjectId = value.substring(lastSep + 1);
        if (unqualifiedObjectId.isEmpty()) {
            throw InvalidObjectIdException.create("empty object ID: " + value);
        }
        return new ObjectId(collectionId, unqualifiedObjectId, value);
    }

    private ObjectId(final CollectionId collectionId, final String unqualifiedObjectId, final String value) {
        super(value);
        this.collectionId = checkNotNull(collectionId);
        this.unqualifiedObjectId = checkNotNull(unqualifiedObjectId);
    }

    public CollectionId getCollectionId() {
        return collectionId;
    }

    public InstitutionId getInstitutionId() {
        return collectionId.getInstitutionId();
    }

    public String getUnqualifiedObjectId() {
        return unqualifiedObjectId;
    }

    private final CollectionId collectionId;
    private final String unqualifiedObjectId;
}

File: py/src/dressdiscover/api/models/object/object_id.py
from dressdiscover.api.models.collection.collection_id import CollectionId
from dressdiscover.api.models.collection.invalid_collection_id_exception import InvalidCollectionIdException
from dressdiscover.api.models.object.invalid_object_id_exception import InvalidObjectIdException


class ObjectId(str):
    @classmethod
    def parse(cls, str_):
        if len(str_) == 0:
            raise InvalidObjectIdException("empty")
        split = str_.rsplit('/', 1)
        if len(split) != 2:
            raise InvalidObjectIdException("malformed object ID: " + str_)
        try:
            collection_id = CollectionId.parse(split[0].strip())
        except InvalidCollectionIdException, e:
            raise InvalidObjectIdException(str(e))
        unqualified_object_id = split[1].strip()
        if len(unqualified_object_id) == 0:
            raise InvalidObjectIdException("empty object id")
        parsed = cls(str_)
        parsed.__collection_id = collection_id
        parsed.__unqualified_object_id = unqualified_object_id
        return parsed

    @property
    def collection_id(self):
        return self.__collection_id

    @property
    def unqualified_object_id(self):
        return self.__unqualified_object_id

